packages/algorand-js/src/interpreter/opcode-list.ts

```diff
@@ -545,13 +578,52 @@ export class Substring3 extends Op {
     const start = this.assertBigInt(stack.pop());
     const end = this.assertBigInt(stack.pop());
 
-    if (end < start) {
-      throw new TealError(ERRORS.TEAL.SUBSTRING_END_BEFORE_START);
-    }
-    if (start > byteString.length || end > byteString.length) {
-      throw new TealError(ERRORS.TEAL.SUBSTRING_RANGE_BEYOND);
-    }
+    const subString = this.subString(start, end, byteString);
+    stack.push(subString);
+  }
+}
+
+/** Pseudo-Ops **/
 
-    stack.push(byteString.slice(Number(start), Number(end)));
+// push integer to stack
+export class Int extends Op {
+  readonly uint64: bigint;
+
+  constructor (uint64: bigint) {
+    super();
+    this.uint64 = uint64;
+  }
+
+  execute (stack: TEALStack): void {
+    this.checkOverflow(this.uint64);
+    stack.push(this.uint64);
+  }
+}
+
+// push bytes to stack
+export class Byte extends Op {
+  readonly byte: Uint8Array;
+
+  constructor (byte: Uint8Array) {
+    super();
+    this.byte = byte;
+  }
+
+  execute (stack: TEALStack): void {
+    stack.push(this.byte);
+  }
+}
+
+// push algorand address as bytes to stack
+export class Addr extends Op {
+  readonly addr: Uint8Array;
+
+  constructor (addr: Uint8Array) {
+    super();
+    this.addr = addr; // parser should verify the addr first
```

Comment:
```
addr parses an Algorand account address base32 and converts it to a regular bytes constant.
```

Reply:
We are already building a parser. In the opcodes we are taking parsed input in the form of `stackElem` (`bigint`/`Uint8array`). Also mentioned this in the comment.

So: Parser will read the string and decode it to algorand byte address and push `Addr(parsed_addr)` to `logic[..]`

Reply:
Yes, addr is not an opcode, I think if parser handles this then we don't need to include pseudo codes because instead of pushing addr(..). Parser will push value to the stack.

But to calculate that value we are supposed to use this pseudo op.
Instead of giving us converted value, it only initializes value. 

I am thinking if we are writing pseudo opcodes, they should take value and return value after doing some operation , according to their respective work instead of initialization because if initialization is only thing they do, we shouldn't push addr(..), we can directly push value to stack.

Reply:
Sorry, i don't follow. Parser does not have access to stack. During parsing the execution has not started. So pushing to stack does not make sense

Reply:
ideally, `addr(string_value)` this should push converted regular bytes to stack. right ?

Reply:
what you are proposing here is, instead of `string_value` , parser will pass bytes value and this function will only push value and not convert it ? am i following you ?

Reply:
TEAL:
```
...
addr JAKNKAMEW63I3H3AXWINXEWW6CCLKX4MQGNTTNBSSUHNCHSLQURMMO47XM
...
```
Parser will read this string, and convert it to bytes (after verifying) and create opcode - Addr(parsed_addr_as_bytes)
During executing, `Addr(..)` will push this value to stack.

Reply:
Correction: Parser will read this string, create opcode - Addr(parsed_addr_as_given)
During executing, Addr(..) will push converted value to stack. 
Parser will parse, but opcodes must do their respective operation
Addr(..): convert it to bytes (after verifying)  and will push this value to stack.

Reply:
yes, if we chose the processing in parser, then we can have a single opcode for `Addr, Int, Bytes`. That opcode will need to take as an argument `elem: StackElem` (bytes or bigint).

Reply:
As per our discussion i have updated the pseudo ops to do the conversion. Parser will do the verification.